### ISSUE TITLE
MockRetrofit and Observable delegate return value is evaluate on api call, not on the subscribe() call

### ISSUE THREAD

**Issue by NitroG42** [square/retrofit#1838]
I will try to explain this problem the best I can. I'm currently using Retrofit, MockRetrofit for the test, and the RxJava integration.
Let's assume we have a networkApi.getCat() that return an Observable<Result<Cat>>.

I want a test that works as following : At the first api call, it will return an error code (401). On the second call, it will return the cat. 

I can simply do something like (sorry it's in kotlin but it won't change the problem) : 

```
object : MockNetworkApi() {
            override fun getCat(): Observable<Result<Cat>> {
                if(firstCall) {
                    firstCall = false
                    delegate.returnFail(401).getCat()
                } else {
                    delegate.returnResponse(Cat()).getCat()
                }
            }
```

So now in my code, if I call the service twice :

```
networkApi.getCat().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe {
    // 401              
} 

networkApi.getCat().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe {
    // cat
} 
```

And it will works as wanted !

Now here's my problem (sorry for the long post). In my production, if I get error 401, I need to use the authenticate api. And I want to do this in the easiest way.
So I have a method startQuery that will log if it gets 401 :

```
fun startQuery(obs: Observable<Result<T>>) { // Obs = networkApi.getCat()
    obs.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe {
        if(result.error == 401) {
            networkApi.authenticate().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe {
                if(result.isSuccess()) {
                    startQuery(obs)
                }
            }
        }
    }   
}
```

But with MockRetrofit, the value returned by the delegate is fixed in the call of networkApi.getCat(), not when we call .subscribe().
So to summarize : only delegate.returnFail(401).getCat()  is returned (because set at the first getCat() call), and calling subscribe twice will only get me a 401 twice.

I understand it's a very specific case, I know the answer is probably "code better, your way is pretty shitty", but I still wanted to post it here.
Thank you for reading through this.

**Comment by JakeWharton** [comment]
I just sent a PR to add `Calls.defer` which will allow this to be supported. Unfortunately we have no good way of calling the actual method more than one. Perhaps we can fix this in Retrofit 3.x, should it ever happen.

The way you'd make this work with `defer` is to check your boolean inside the `Callable` and then return either `Calls.response` with either the success value or the 401.
